include/pci.h

```diff
@@ -81,32 +84,28 @@ struct pci_device {
 
   unsigned nbars;
   resource_t bar[6];
-};
+} pci_dev_data_t;
 
-/* TODO: pci_bus_device will become a state (device_t) of generic PCI driver. */
-typedef struct pci_bus_device {
-  pci_bus_t *bus;
-  resource_t *mem_space;
-  resource_t *io_space;
-  pci_device_list_t devices;
-} pci_bus_device_t;
+#define PCI_DRIVER(dev) ((pci_bus_driver_t *)((dev)->parent->driver))
 
-static inline uint32_t pci_read_config(pci_device_t *device, unsigned reg,
+static inline uint32_t pci_read_config(device_t *device, unsigned reg,
                                        unsigned size) {
-  return device->bus->read_config(device, reg, size);
+  return PCI_DRIVER(device)->pci_bus.read_config(device, reg, size);
 }
 
-static inline void pci_write_config(pci_device_t *device, unsigned reg,
+static inline void pci_write_config(device_t *device, unsigned reg,
                                     unsigned size, uint32_t value) {
-  device->bus->write_config(device, reg, size, value);
+  PCI_DRIVER(device)->pci_bus.write_config(device, reg, size, value);
 }
 
-static inline uint32_t pci_adjust_config(pci_device_t *device, unsigned reg,
+static inline uint32_t pci_adjust_config(device_t *device, unsigned reg,
                                          unsigned size, uint32_t value) {
   pci_write_config(device, reg, size, value);
   return pci_read_config(device, reg, size);
 }
 
-void pci_init();
+void pci_bus_enumerate(device_t *pcib);
+void pci_bus_assign_space(device_t *pcib);
+void pci_bus_dump(device_t *pcib);
```

Comment:
Don't you think this entire header would be much cleaner if all these functions took `pci_device_t*` as argument?

Also, I see that all these functions blindly assume that the `device_t` they are given is indeed a PCI device. I hope this will be always the case, but think what horrible problems will happen, if, due to a bug or a mistake, they are passed a different kind of device. If we really don't want to use `pci_device_t*` type and let the compiler check this for us, then I suggest to introduce an enum inside `device_t` which would store device type, so that we could `assert(device->type == DEV_TYPE_PCI);`.